README.markdown
# puppet_sslforfree #

[Português]: #português
[Requisitos]: #requisitos
[Instruções]: #instruções-de-uso
[Hiera]: #hiera
[English]: #english
[Requirements]: #requirements
[Instructions]: #instructions
[Hiera]: #hiera
[Parameters]: #parameters
[Developers]: #developers
[License]: #license

#### Menu

1. [Português][Português]
    - [Requisitos][requisitos]
    - [Instruções de uso][Instruções]
    - [Hiera][Hiera]
    - [TASK_README.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/TASK_README.md)
2. [English][English]
    - [Requirements][requirements]
    - [Instructions][instructions]
    - [Hiera][Hiera]
    - [TASK_README.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/TASK_README.md)
3. [Parameters][Parameters]
4. [Developers][Developers]
5. [CHANGELOG.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/CHANGELOG.md)
6. [License][License]

# Português

Este é o modulo *puppet_sslforfree*.

Instala e gerencia o certificado assinado pelos sites:

- https://www.sslforfree.com
- https://zerossl.com
- https://gethttpsforfree.com

Também cria o certificado no formato JKS.

Para aprender mais sobre o uso do Puppet, recomendo a leitura do livro que
publiquei pela Novatec:

https://novatec.com.br/livros/puppet

Veja também os links que estão nesta página:

http://blog.aeciopires.com/primeiros-passos-com-o-puppet

## Requisitos

1. Acesse um dos sites anteriores e gere o certificado de host ou wildcard conforme
as instruções disponíveis. Ambos usam a API do Let's encrypt para assinar os
certificados válidos e gratuitos com duração de 90 dias.
2. Puppet 4.x ou superior
3. Sistema operacional: Debian 8.x, 9.x, CentOS 6.x, 7.x, Red Hat 6.x e
7.x, Ubuntu 14.04, 16.04 e 18.04.
4. Instalar o pacote ``keytool`` obtido junto com o Java. Este módulo não
instala o Java.

Observações:

- Para atualizar/renovar o certificado basta mudar o valor do parâmetro
``overwrite_certificate`` para ``true`` e informar a nova URL de download em
``cert_download_url_base``, após gerar o novo certificado e salvá-lo em um
servidor web.
- Este módulo não configura os serviços para usar o certificado. Isso
deve ser feito por outro módulo Puppet ou manualmente.

## Instruções de Uso

Baixe o módulo em:
**https://github.com/aeciopires/puppet_sslforfree/releases**

Descompacte o pacote e copiar o diretório **puppet_sslforfree** para a
máquina **puppetserver**.

Na máquina **puppetserver**, mova o diretório **puppet_sslforfree**
para o diretório de módulos, por exemplo:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/modules/**. Onde ``NAME_ENVIRONMENT`` deve ser trocado pelo nome do environment que você
quer usar no PuppetServer.

Edite o aquivo
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/manifests/site.pp**
e definir quais hosts usarão o módulo, conforme o exemplo abaixo. Exemplo
 da configuração do arquivo ``site.pp``.

~~~ puppet
node node1.domain.com.br {
    include puppet_sslforfree
}
~~~

Execute o Puppet Agent no servidor ``node1.domain.com.br``.

~~~ bash
puppet agent -t
~~~

## Hiera

O módulo *puppet_sslforfree* instala e configura o certificado com as
configurações definidas em parâmetros ou variáveis declaradas no
manifest **params.pp**.

Algumas variáveis possuem valores customizados de acordo com o servidor.
Estes valores são obtidos através do Hiera (com dados armazenados em
arquivos do tipo "*.yaml").

Abaixo está um exemplo do arquivo de configuração do Hiera, que deve
ficar localizado em:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hiera.yaml**

~~~ puppet
---
version: 5
defaults:
  datadir: hieradata
  data_hash: yaml_data
hierarchy:
  - name: "Hosts"
    paths:
      - "host/%{::trusted.certname}.yaml"
      - "host/%{::facts.networking.fqdn}.yaml"
      - "host/%{::facts.networking.hostname}.yaml"

  - name: "Dominios"
    paths:
      - "domain/%{::trusted.domain}.yaml"
      - "domain/%{::domain}.yaml"

  - name: "Dados comuns"
    path: "common.yaml"
~~~

Dessa forma, o Hiera buscará, prioritariamente, os valores definidos nas
variáveis de host (sobrepondo os valores de variáveis de mesmo nome
definidas por domínio). Estas variáveis devem ficar em arquivos como esse:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/host/node1.domain.com.br.yaml**.

Caso não sejam definidos valores para variáveis nos arquivos de hosts,
o Hiera buscará valores definidos em variáveis de domínio. As variáveis
de domínio devem ficar em arquivos como esse:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/domain/domain.com.br.yaml**.

As variáveis definidas no arquivo
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/common.yaml**,
só serão aplicadas em último caso.

Mesmo que nenhum destes arquivos existam, serão aplicados os valores padrão
definidos na classe **params.pp**.

### Exemplo do arquivo .yaml

~~~ puppet
#---------------------
#BEGIN
#---------------------

#Espaco requerido: 2 MB ou 2.000.000 bytes
space_required: 2000000
tmp_dir: '/tmp'
manage_certificate_jks: true
overwrite_certificate: false
download_certificate: true
cert_download_url_base: 'https://192.168.0.1/cert'
keytool: '/usr/bin/keytool'
host_cert_key: 'private.key'
host_cert_crt: 'certificate.crt'
ca_cert: 'ca_bundle.crt'
host_cert_pass: ''
cert_alias: 'sslfree'
ca_cert_alias: 'ca_sslfree'
certs_dir: '/etc/sslfree'
java_cacert: '/usr/lib/jvm/java-8-oracle/jre/lib/security/cacerts'

#---------------------
#END
#---------------------
~~~

# English

This is the **puppet_sslforfree** module.

Installs and manages the certificate signed by the sites:

- https://www.sslforfree.com
- https://zerossl.com
- https://gethttpsforfree.com

It also creates the certificate in JKS format.

## Requirements

1. Access one of the sites previous and generate the host or wildcard certificate.
Both use the Let's encrypt API to sign the valid and free certificates with a
duration of 90 days.
2. Puppet 4.x or higher
3. Operating System: Debian 8.x, 9.x, CentOS 6.x, 7.x, Red Hat 6.x and
7.x, Ubuntu 14.04, 16.04 and 18.04.
4. Install the ``keytool`` packages obtained with Java. This module does not
installs Java.

Comments:

- To update/renew the certificate, simply change the value of the parameter
``overwrite_certificate`` to `` true`` and enter the new download URL at
``cert_download_url_base`` after generating the new certificate and save it to
a web server.
- This module does not configure services to use the certificate. You
MUST be done by another Puppet module or manually.

## Instructions

Download the module in:
**https://github.com/aeciopires/puppet_sslforfree/releases**

Unzip the package and copy the **puppet_sslforfree** directory to the
 **puppetserver** machine.

On the **puppetserver** machine, move the directory **puppet_sslforfree**
to the modules directory, for example:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/modules/**. Where ``NAME_ENVIRONMENT`` should be changed by the name of the environment
you want to use in PuppetServer.

Edit the file
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/manifests/site.pp**
and define which hosts will use the module, as shown in the example below.
Example of the configuration of the ``site.pp`` file

~~~ puppet
node node1.domain.com.br {
    include puppet_sslforfree
}
~~~

Run the Puppet Agent on the ``node1.domain.com`` server.

~~~ bash
puppet agent -t
~~~

## Hiera

The **puppet_sslforfree** module installs and configures the certificate with
the settings defined in parameters or variables declared in the manifest
**params.pp**.

Some variables have custom values ​​according to the server.
These values ​​are obtained through Hiera (with data stored in
files of type "* .yaml").

Below is an example of the Hiera configuration file, which should
be located at:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hiera.yaml**

~~~ puppet
---
version: 5
defaults:
  datadir: hieradata
  data_hash: yaml_data
hierarchy:
  - name: "Hosts"
    paths:
      - "host/%{::trusted.certname}.yaml"
      - "host/%{::facts.networking.fqdn}.yaml"
      - "host/%{::facts.networking.hostname}.yaml"

  - name: "Domains"
    paths:
      - "domain/%{::trusted.domain}.yaml"
      - "domain/%{::domain}.yaml"

  - name: "Common"
    path: "common.yaml"
~~~

In this way, Hiera will seek, as a priority, the values ​​defined in the
variables (overlapping variable values ​​of the same name defined by the domain).
These variables should be in files like this:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/host/node1.domain.com.br.yaml**.

If no value is set for variables in the hosts files,
Hiera will look for values ​​defined in domain variables. The variables
domain MUST be in files like this:
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/domain/domain.com.br.yaml**.

The variables defined in the
**/etc/puppetlabs/code/environments/NAME_ENVIRONMENT/hieradata/common.yaml**
file, will only be applied in the latter case.

Even if none of these files exist, the default values ​​defined in the
**params.pp** class will be applied.

### Sample file .yaml

~~~ puppet
#---------------------
#BEGIN
#---------------------

#Space Required: 2 MB or 2.000.000 bytes
space_required: 2000000
tmp_dir: '/tmp'
manage_certificate_jks: true
overwrite_certificate: false
download_certificate: true
cert_download_url_base: 'https://192.168.0.1/cert'
keytool: '/usr/bin/keytool'
host_cert_key: 'private.key'
host_cert_crt: 'certificate.crt'
ca_cert: 'ca_bundle.crt'
host_cert_pass: ''7
cert_alias: 'sslfree'
ca_cert_alias: 'ca_sslfree'
certs_dir: '/etc/sslfree'
java_cacert: '/usr/lib/jvm/java-8-oracle/jre/lib/security/cacerts'

#---------------------
#END
#---------------------
~~~

## Parameters

### space_required

**Description:** Sets the minimum required disk space.

**Data type:** Interger.

**Default value:** 2000000 (in bytes).

### tmp_dir

**Description:** Sets the temporary directory.

**Data type:** String.

**Default value:** /tmp

### manage_certificate_jks

**Description:** If ``true``, it manages the certificate. If ``false`` does
nothing.

**Data type:** Boolean.

**Default value:** true

### overwrite_certificate

**Description:** If ``true``, overwrite the certificate with the same alias
every round of the puppet agent. If ``false``, keep the certificate and to
register the new you must change the alias of the new certificate in the
parameters: ``cert_alias`` and ``ca_cert_alias``.

**Data type:** Boolean.

**Default value:** false

### download_certificate

**Description:** If ``true``, download the certificate from the URL formed by
the concatenation of the parameters: ``$cert_download_url_base/$host_cert_key``,
``$cert_download_url_base/$host_cert_crt`` and
``$cert_download_url_base/$ca_cert``. If ``false``, you must to update the
content of the files in the module directory in ``files/certs_sslfree``.

**Data type:** Boolean.

**Default value:** true

### cert_download_url_base

**Description:** Certificate download base URL.

**Data type:** String.

**Default value:** https://192.168.0.1/cert

### keytool

**Description:** Path of the keytool binary in the operating system. Keytool is
obtained by installing Java. This module does not manage Java installation

**Data type:** String.

**Default value:** /usr/bin/keytool

### host_cert_key

**Description:** Name of the private key file. The name of this file will be
concatenated with the URL entered in the parameter ``cert_download_url_base``.

**Data type:** String.

**Default value:** private.key

### host_cert_crt

**Description:** Name of the certificate file. The name of this file will be
concatenated with the URL entered in the parameter ``cert_download_url_base``.

**Data type:** String.

**Default value:** certificate.crt

### ca_cert

**Description:** Name of the certificate file of the Certificate Authority.
The name of this file will be concatenated with the URL entered in the
parameter ``cert_download_url_base``.

**Data type:** String.

**Default value:** ca_bundle.crt

### host_cert_pass

**Description:** Certificate password.

**Data type:** String.

**Default value:** ' ' (empty)

### cert_alias

**Description:** Certificate alias. This module uses the keytool to create a
repository JKS so that the certificate can be used by applications or
Java application servers.

**Data type:** String.

**Default value:** sslfree

### ca_cert_alias

**Description:** Alias of the Certificate Authority. This module uses the
keytool to create a repository JKS so that the certificate can be used
by applications or Java application servers.

**Data type:** String.

**Default value:** ca_sslfree

### certs_dir

**Description:** Directory in which the certificates will be stored.

**Data type:** String.

**Default value:** /etc/sslfree

### java_cacert

**Description:** Repository JKS of the Java for storing Certificate Authority
certificates. See https://en.wikipedia.org/wiki/Keystore.

**Data type:** String.

**Default value:** /usr/lib/jvm/java-8-oracle/jre/lib/security/cacerts

## Developers

developer: Aecio dos Santos Pires<br>
mail: aeciopires at gmail.com<br>

## License

Apache 2.0 2018 Aécio dos Santos Pires


CHANGELOG.md
# Change log

All notable changes to this project will be documented in this file. The format
is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org).

## 2018-04-30 Release [1.3.1]
### Summary

This is a patch release, it includes documentation update.

## 2018-04-29 Release [1.3.0]
### Summary
This is a minor release, it includes enhancements to the documentation and
support to https://zerossl.com and https://gethttpsforfree.com.

#### Added
- Parameter overwrite_certificate. See
[README.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/README.markdown)
- Updated documentation

## 2018-04-28 Release [1.2.0]
### Summary
This is a minor release, it includes enhancements to the documentation and new
configuration parameter and code snippet that can replace an existing
certificate in a JKS repository.

#### Added
- Parameter overwrite_certificate. See
[README.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/README.markdown)
- Updated documentation

## 2018-04-24 Release [1.1.0]
### Summary
This is a minor release, it makes this module PDK-compliant for easier
maintenance and includes task for execution with Puppet-Bolt.

#### Added
- PDK conversion.
- Tasks. See
[TASK_README.md](https://github.com/aeciopires/puppet_sslforfree/blob/master/TASK_README.md)
- Updated documentation

## 2018-04-23 Release [1.0.2]
### Summary

This is a pacth release, it includes the parameter to support the path of
java_cacerts.

#### Added
- Parameter 'java_cacert'.
- Support added for Ubuntu 18.04.

## 2018-04-23 Release 1.0.1
### Summary
Bad release. Error in documentation.

## 2018-04-23 Release 1.0.0
### Summary

Publish module. Hello Word! :-)


spec/spec_helper.rb
require 'puppetlabs_spec_helper/module_spec_helper'
require 'rspec-puppet-facts'

begin
  require 'spec_helper_local' if File.file?(File.join(File.dirname(__FILE__), 'spec_helper_local.rb'))
rescue LoadError => loaderror
  warn "Could not require spec_helper_local: #{loaderror.message}"
end

include RspecPuppetFacts

default_facts = {
  puppetversion: Puppet.version,
  facterversion: Facter.version
}

default_facts_path = File.expand_path(File.join(File.dirname(__FILE__), 'default_facts.yml'))
default_module_facts_path = File.expand_path(File.join(File.dirname(__FILE__), 'default_module_facts.yml'))

if File.exist?(default_facts_path) && File.readable?(default_facts_path)
  default_facts.merge!(YAML.safe_load(File.read(default_facts_path)))
end

if File.exist?(default_module_facts_path) && File.readable?(default_module_facts_path)
  default_facts.merge!(YAML.safe_load(File.read(default_module_facts_path)))
end

RSpec.configure do |c|
  c.default_facts = default_facts
end


tasks/init.sh
#!/bin/bash

#-------------------------------------------------------
# file:         init.sh
# comment:      Install and manage the certificate signed by the sites
# https://www.sslforfree.com
# https://zerossl.com
# https://gethttpsforfree.com
# Also creates the certificate in JKS format.
# author:       Aecio Pires <aeciopires@gmail.com>
# date:         24-abr-2018
# revision:     Aecio Pires <aeciopires@gmail.com>
# last updated: 30-abr-2018, 09:47
#-------------------------------------------------------

#By default tasks take arguments as environment variables, prefixed with PT
# (short for Puppet Tasks)
# https://github.com/puppetlabs/tasks-hands-on-lab/tree/master/5-writing-tasks

#----------------------------------------------------
function isroot(){
	MYUID=$(id | cut -d= -f2 | cut -d\( -f1)
	[ "$MYUID" -eq 0 ] && echo YES && return 0 || echo NO && return 1
}

#------------------------------------------------------------
function install_essential_packages(){
PACKAGES="$1"
echo "[INFO] Installing necessary packages..."
case "$LINUXDISTRO" in
  DEBIAN|UBUNTU)
    sudo apt-get update > /dev/null 2>&1
    sudo apt-get install -y $PACKAGES > /dev/null 2>&1
  ;;
  CENTOS|REDHAT)
    yum -y install $PACKAGES > /dev/null 2>&1
  ;;
  *)
    echo "[ERROR] I don't know how to install packages [$PACKAGES] in this \
          distro: $LINUXDISTRO." && exit 3
    ;;
  esac
}

#------------------------------------------------------------
function check_essential_commands(){

PACKAGES="$1"

echo "[INFO] Checking in the system the essential commands for the script..."
for file in "$PACKAGES" ; do
  if [ ! -f "$file" ] ; then
    echo "[ERROR] File not found: \"$file\""
    return 3
  fi
done

return 0
}

#-----------------------------------------------------------
function get_linux_distro(){

LINUXDISTRO=UNKNOWN

grep "CentOS" /etc/redhat-release > /dev/null 2>&1 && LINUXDISTRO=CENTOS
grep "Red Hat" /etc/redhat-release > /dev/null 2>&1 && LINUXDISTRO=REDHAT
grep "Debian" /etc/issue > /dev/null 2>&1 && LINUXDISTRO=DEBIAN
grep "Ubuntu" /etc/issue > /dev/null 2>&1 && LINUXDISTRO=UBUNTU

echo $LINUXDISTRO
}

#------------------------------------------------------------
function test_distro_suport(){

echo "[INFO] Testing script compatibility with distro..."

case "$LINUXDISTRO" in
  CENTOS|REDHAT|DEBIAN|UBUNTU)
    echo "[INFO] Support OK: $LINUXDISTRO"
  ;;
  *)
    echo "[ERROR] This script is not ready to run in the distro $LINUXDISTRO."
    exit 5
  ;;
esac
}

#-----------------------------------------------
#-----------------------------------------------
# MAIN
#-----------------------------------------------
#-----------------------------------------------

# Declaracao de variaveis
PACKAGES="wget openssl keytool"
BINARIES="/usr/bin/wget /usr/bin/openssl /usr/bin/keytool"
CMDLINE=$(readlink --canonicalize --no-newline "$BASH_SOURCE")
PROGFILENAME=$(basename "$BASH_SOURCE")
PROGDIRNAME=$(dirname $(readlink -f "$BASH_SOURCE"))
LINUXDISTRO=$(get_linux_distro)
DATA=`date +%Y%m%d_%H-%M`

#Variaveis defaults caso estejam vazias
PT_certs_dir=${PT_certs_dir:-'/etc/sslfree'}
PT_keystore_file=${PT_keystore_file:-'/etc/sslfree/keystore.jks'}
PT_cacerts_file=${PT_cacerts_file:-'/etc/sslfree/cacerts.jks'}
PT_download_ca_cert=${PT_download_ca_cert:-'https://192.168.0.1/cert/ca_bundle.crt'}
PT_ca_cert=${PT_ca_cert:-'ca_bundle.crt'}
PT_download_host_cert_key=${PT_download_host_cert_key:-'https://192.168.0.1/cert/private.key'}
PT_host_cert_key=${PT_host_cert_key:-'private.key'}
PT_download_host_cert_crt=${PT_download_host_cert_crt:-'https://192.168.0.1/cert/certificate.crt'}
PT_host_cert_crt=${PT_host_cert_crt:-'certificate.crt'}
PT_cert_alias=${PT_cert_alias:-'sslfree'}
PT_host_cert_pass=${PT_host_cert_pass:-''}
PT_ca_cert_alias=${PT_ca_cert_alias:-'ca_sslfree'}
PT_java_cacert=${PT_java_cacert:-'/usr/lib/jvm/java-8-oracle/jre/lib/security/cacerts'}
PT_keystore_pass_default=${PT_keystore_pass_default:-'changeit'}
PT_keystore_pass=${PT_keystore_pass:-'changeit'}
#------------------------------------------------------------

# Detecta se eh um usuario com poderes de root que esta executando o script
#if [ $(isroot) = NO ] ; then
#  echo "voce deve ser root para executar este script."
#  echo "execute o comando \"sudo $CMDLINE\""
#  exit 4
#fi

# Detecta se a distro GNU/Linux eh suportada
test_distro_suport

# Checa os comandos essenciais a execucao do script
if ! check_essential_commands $BINARIES ; then
  install_essential_packages $PACKAGES
  check_essential_commands $BINARIES|| exit 3
fi

# Backp do diretorio de certificados
sudo mkdir -p "$PT_certs_dir"
sudo cp -R "$PT_certs_dir" "$PT_certs_dir"-backup-$DATA;

# Download dos certificados
echo "[INFO] Downloading the certificates..."

sudo wget --no-check-certificate -q "$PT_download_ca_cert" \
  -O "$PT_certs_dir/$PT_ca_cert"

sudo wget --no-check-certificate -q "$PT_download_host_cert_key" \
  -O "$PT_certs_dir/$PT_host_cert_key"

sudo wget --no-check-certificate -q "$PT_download_host_cert_crt" \
  -O "$PT_certs_dir/$PT_host_cert_crt"

echo "[INFO] Verifying that the certificates exist..."

sudo chmod -R 755 "$PT_certs_dir"

if [ -f "$PT_certs_dir/$PT_ca_cert" ] ; then
  if [ -z "$PT_certs_dir/$PT_ca_cert" ] ; then
    echo "[ERROR] File empty: '$PT_certs_dir/$PT_ca_cert'"
    exit 15
  fi
else
  echo "[ERROR] File not found: '$PT_certs_dir/$PT_ca_cert'"
  exit 14
fi

if [ -f "$PT_certs_dir/$PT_host_cert_key" ] ; then
  if [ -z "$PT_certs_dir/$PT_host_cert_key" ] ; then
    echo "[ERROR] File empty: '$PT_certs_dir/$PT_host_cert_key'"
    exit 17
  fi
else
  echo "[ERROR] File not found: '$PT_certs_dir/$PT_host_cert_key'"
  exit 16
fi

if [ -f "$PT_certs_dir/$PT_host_cert_crt" ] ; then
  if [ -z "$PT_certs_dir/$PT_host_cert_crt" ] ; then
    echo "[ERROR] File empty: '$PT_certs_dir/$PT_host_cert_crt'"
    exit 19
  fi
else
  echo "[ERROR] File not found: '$PT_certs_dir/$PT_host_cert_crt'"
  exit 18
fi

if [ -f "$PT_keystore_file" ] ; then
  if [ ! -z "$PT_keystore_file" ] ; then
    # Removendo o certificado caso ja exista
    if sudo keytool -list -keystore "$PT_keystore_file" -alias "$PT_cert_alias" \
      -v -storepass "$PT_keystore_pass" -noprompt > /dev/null 2>&1; then

        sudo keytool -delete -keystore "$PT_keystore_file" -alias "$PT_cert_alias" \
          -storepass "$PT_keystore_pass" -noprompt > /dev/null 2>&1;
    fi
  else
    echo "[ERROR] File empty: '$PT_keystore_file'"
    exit 7
  fi
#else
#  echo "[ERROR] File not found: '$PT_keystore_file'"
#  exit 6
fi

echo "[INFO] Deploying the certificates in '$PT_certs_dir'..."

# Adicionando o certificado no keystore
sudo openssl pkcs12 -export -name "$PT_cert_alias" -in "$PT_certs_dir/$PT_host_cert_crt" \
  -inkey "$PT_certs_dir/$PT_host_cert_key" -out "$PT_certs_dir/keystore.p12" -password pass:"$PT_host_cert_pass"

sudo keytool -importkeystore -destkeystore "$PT_keystore_file" \
  -srckeystore "$PT_certs_dir/keystore.p12" -srcstoretype pkcs12 -alias "$PT_cert_alias" \
  -deststorepass "$PT_keystore_pass" -srcstorepass "$PT_host_cert_pass"

# Alterando a senha da chave privada para ser igual a do keystore.
sudo keytool -keypasswd -keypass "$PT_host_cert_pass" -alias "$PT_cert_alias" \
  -keystore "$PT_keystore_file" -storepass "$PT_keystore_pass" \
  -new "$PT_keystore_pass"

if [ -f "$PT_cacerts_file" ] ; then
  if [ ! -z "$PT_cacerts_file" ] ; then
    # Removendo o certificado caso ja exista
    if sudo keytool -list -keystore "$PT_cacerts_file" -alias "$PT_ca_cert_alias" \
      -v -storepass "$PT_keystore_pass" -noprompt > /dev/null 2>&1; then

        sudo keytool -delete -keystore "$PT_cacerts_file" -alias "$PT_ca_cert_alias" \
          -storepass "$PT_keystore_pass" -noprompt > /dev/null 2>&1
    fi
  else
    echo "[ERROR] File empty: '$PT_cacerts_file'"
    exit 9
  fi
#else
#  echo "[ERROR] File not found: '$PT_cacerts_file'"
#  exit 8
fi

echo "[INFO] Deploying the CA certificates in '$PT_certs_dir'..."

# Adicionando o certificado do CA no cacerts
sudo keytool -import -v -trustcacerts -alias "$PT_ca_cert_alias" -file "$PT_certs_dir/$PT_ca_cert" \
  -keystore "$PT_cacerts_file" -storepass "$PT_keystore_pass" -noprompt

# Adicionando o certificado do trustore do Java
sudo cp -R "$PT_java_cacert" "$PT_java_cacert"-backup-$DATA

if [ -f "$PT_java_cacert" ] ; then
  if [ ! -z "$PT_java_cacert" ] ; then
    # Removendo o certificado caso ja exista
    if sudo keytool -list -keystore "$PT_java_cacert" -alias "$PT_ca_cert_alias" \
      -v -storepass "$PT_keystore_pass_default" -noprompt > /dev/null 2>&1; then

        sudo keytool -delete -keystore "$PT_java_cacert" -alias "$PT_ca_cert_alias" \
          -storepass "$PT_keystore_pass_default" -noprompt > /dev/null 2>&1
    fi
  else
    echo "[ERROR] File empty: '$PT_java_cacert'"
    exit 11
  fi
else
  echo "[ERROR] File not found: '$PT_java_cacert'"
  exit 10
fi

echo "[INFO] Deploying the CA certificate in '$PT_java_cacert'..."
sudo keytool -import -v -trustcacerts -alias "$PT_ca_cert_alias" \
  -file "$PT_certs_dir/$PT_ca_cert" -keystore "$PT_java_cacert" \
  -storepass "$PT_keystore_pass_default" -noprompt

if [ -f "$PT_java_cacert" ] ; then
  if [ ! -z "$PT_java_cacert" ] ; then
    # Removendo o certificado caso ja exista
    if sudo keytool -list -keystore "$PT_java_cacert" -alias "$PT_cert_alias" \
      -v -storepass "$PT_keystore_pass_default" -noprompt > /dev/null 2>&1; then

        sudo keytool -delete -keystore "$PT_java_cacert" -alias "$PT_cert_alias" \
          -storepass "$PT_keystore_pass_default" -noprompt > /dev/null 2>&1
    fi
	else
    echo "[ERROR] File empty: '$PT_java_cacert'"
    exit 11
  fi
else
  echo "[ERROR] File not found: '$PT_java_cacert'"
  exit 10
fi

echo "[INFO] Deploying the certificate in '$PT_java_cacert'..."

sudo keytool -import -v -trustcacerts -alias "$PT_cert_alias" \
  -file "$PT_certs_dir/$PT_host_cert_crt" -keystore "$PT_java_cacert" \
  -storepass "$PT_keystore_pass_default" -noprompt

# Ajustando as permissoes de acesso
echo "[INFO] Setting permissions for '$PT_certs_dir' directory access..."

sudo chown -R root:root "$PT_certs_dir"
sudo chmod -R 755 "$PT_certs_dir"
sudo chmod -R 750 "$PT_certs_dir/$PT_host_cert_key";
